File: Sum of Prefix Scores of Strings.cpp
class Node {
public:
    unordered_map<char, Node> route;
    int score = 0;
    Node() {};
    void add(string& word, int idx) {
        if (word.size() == idx) {
            return;
        }

        auto c = word[idx];
        if (!route.contains(c)) route[c] = Node();
        route[c].score++;
        route[c].add(word, idx+1);
    }

    int find(string& word, int idx) {
        if (word.size() == idx) return score;

        auto c = word[idx];
        if (!route.contains(c)) return score;
        return score + route[c].find(word, idx+1);
    }
};

class Solution {
public:
    vector<int> sumPrefixScores(vector<string>& words) {
        auto trie = Node();
        for (auto word: words) {
            trie.add(word, 0);
        }

        vector<int> arr;
        arr.reserve(words.size());
        for (auto word: words) {
            arr.push_back(trie.find(word, 0));
        }
        return arr;
    }
};

File: Sum of Prefix Scores of Strings.py
class Node:
    def __init__(self):
        self.route = {}
        self.score = 0

    def add(self, word, idx):
        if len(word) == idx:
            return

        c = word[idx]
        if c not in self.route:
            self.route[c] = Node()
        self.route[c].score += 1
        self.route[c].add(word, idx+1)

    def find(self, word, idx):
        if len(word) == idx:
            return self.score

        c = word[idx]
        if c not in self.route:
            return self.score
        return self.score + self.route[c].find(word, idx+1)

class Solution:
    def sumPrefixScores(self, words: List[str]) -> List[int]:
        trie = Node()
        for word in words:
            trie.add(word, 0)

        arr = []
        for word in words:
            arr.append(trie.find(word, 0))
        return arr
